src/hotspot/share/opto/mulnode.cpp

```diff
@@ -2052,99 +2057,70 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
-      if (phase->type(shift)->isa_integer(bt) == nullptr) {
-        return false;
-      }
-    }
-  }
-  if (shift->Opcode() != Op_LShift(shift_bt)) {
-    if (check_reverse &&
-        (mask->Opcode() == Op_LShift(bt) ||
-         (bt == T_LONG && mask->Opcode() == Op_ConvI2L))) {
-      // try it the other way around
-      return AndIL_shift_and_mask_is_always_zero(phase, mask, shift, bt, false);
-    }
-    return false;
-  }
-  Node* shift2 = shift->in(2);
-  if (shift2 == nullptr) {
-    return false;
-  }
-  const Type* shift2_t = phase->type(shift2);
-  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
-    return false;
+    type = phase->type(expr)->isa_int();
   }
 
-  jint shift_con = shift2_t->is_int()->get_con() & ((shift_bt == T_INT ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
-  if ((((jlong)1) << shift_con) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0) {
-    return true;
+  if (expr->Opcode() == Op_LShift(bt)) {
+    const TypeInt* rhs_t = phase->type(expr->in(2))->isa_int();
+    if (rhs_t == nullptr || !rhs_t->is_con()) {
+      return 0;
+    }
+    return rhs_t->get_con() % (type2aelembytes(bt) * BitsPerByte);
   }
 
-  return false;
+  return 0;
 }
 
-// Given an expression (AndX (AddX v1 (LShiftX v2 #N)) #M)
-// determine if the AndX must always produce (AndX v1 #M),
-// because the shift (v2<<N) is bitwise disjoint from the mask #M.
-// The X in AndX will be I or L, depending on bt.
-// Specifically, the following cases fold,
+// Given an expression (AndX T+expr mask), determine
+// whether expr is neutral wrt addition under mask
+// and hence the result is always equivalent to (AndX T mask),
+// The X in AndX must be I or L, depending on bt.
+// Specifically, this holds for the following cases,
 // when the shift value N is large enough to zero out
-// all the set positions of the and-mask M.
-//   (AndI (AddI v1 (LShiftI _ #N)) #M) => (AndI v1 #M)
-//   (AndL (AddI v1 (LShiftL _ #N)) #M) => (AndL v1 #M)
-//   (AndL (AddL v1 (ConvI2L (LShiftI _ #N))) #M) => (AndL v1 #M)
+// all the set positions of the and-mask M:
+//   (AndI (LShiftI _ #N) #M)
+//   (AndL (LShiftL _ #N) #M)
+//   (AndL (ConvI2L (LShiftI _ #N)) #M)
+// including constant operands:
+//   (AndI (ConI (_ << #N)) #M)
+//   (AndL (ConL (_ << #N)) #M)
 // The M and N values must satisfy ((-1 << N) & M) == 0.
-// Because the optimization might work for a non-constant
-// mask M, and because the AddX operands can come in either
-// order, we check for every operand order.
-Node* MulNode::AndIL_add_shift_and_mask(PhaseGVN* phase, BasicType bt) {
+static bool AndIL_is_zero_element(const PhaseGVN* phase, const Node* expr, const Node* mask, BasicType bt) {
+  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
+  if (mask_t == nullptr) {
+    return false;
+  }
+
+  jint zeros = AndIL_min_trailing_zeros(phase, expr, bt);
+  return zeros > 0 && ((((jlong)1) << zeros) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0);
```

Comment:
This line indicates that the `mask` could be a variable. You should make sure to add some tests for that in your patterns. You can create a variable in a specific range like this `Math.min(5, Math.max(1, x))`, should get you `x` clamped into the region `1..5`.